docs/orleans/tutorials-and-samples/custom-grain-storage.md

```diff
@@ -1,25 +1,163 @@
+    {
+        throw new NotImplementedException();
+    }
+
+    public void Participate(ISiloLifecycle lifecycle) =>
+        throw new NotImplementedException();
+}
+```
+
+Before starting the implementation, you'll create an options class containing the root directory where the grain state files are persisted. For that you'll create an options file named `FileGrainStorageOptions` containing the following:
+
+:::code source="snippets/custom-grain-storage/FileGrainStorageOptions.cs":::
+
+With the options class created, you'll explore the constructor parameters of the `FileGrainStorage` class:
+
+- `storageName`: to specify which grains should write using this storage, for example `[StorageProvider(ProviderName = "File")]`.
+- `options`: the options class we just created.
+- `clusterOptions`: the cluster options used for retrieving the `ServiceId`.
+
+## Initialize the storage
+
+To initialize the storage, you subscribe to the <xref:Orleans.ServiceLifecycleStage.ApplicationServices?displayProperty=nameWithType> stage with an `onStart` function. Consider the following <xref:Orleans.ILifecycleParticipant%601.Participate%2A?displayProperty=nameWithType> implementation:
+
+:::code source="snippets/custom-grain-storage/FileGrainStorage.cs" id="participate":::
+
+The `onStart` function is used to conditionally create the root directory to store the grains states when it doesn't already exist.
+
+You'll also provide a common function to construct the filename ensuring uniqueness per service, grain id, and grain type:
+
+:::code source="snippets/custom-grain-storage/FileGrainStorage.cs" id="getkeystring":::
+
+## Read state
+
+To read a grain state, you get the filename using the `GetKeyString` function and combine it with the root directory coming from the `_options` instance.
+
+:::code source="snippets/custom-grain-storage/FileGrainStorage.cs" id="readstateasync":::
+
+We use the `fileInfo.LastWriteTimeUtc` as an `ETag` which will be used by other functions for inconsistency checks to prevent data loss.
+
+For the deserialization, you use the <xref:Orleans.Storage.IStorageProviderSerializerOptions.GrainStorageSerializer?displayProperty=nameWithType>. This is important to be able to serialize/deserialize properly the state.
+
+## Write state
+
+Writing the state is similar to reading the state.
+
+:::code source="snippets/custom-grain-storage/FileGrainStorage.cs" id="writestateasync":::
+
+Similar to reading state, you use the <xref:Orleans.Storage.IStorageProviderSerializerOptions.GrainStorageSerializer?displayProperty=nameWithType> to write the state. The current `ETag` is used to check against the last updated time in the UTC of the file. If the date is different, it means that another activation of the same grain changed the state concurrently. In this situation, you'll throw an `InconsistentStateException` which will result in the current activation being killed to prevent overwriting the state previously saved by the other activated grain.
```

Comment:
```suggestion
Similar to reading state, you use the <xref:Orleans.Storage.IStorageProviderSerializerOptions.GrainStorageSerializer?displayProperty=nameWithType> to write the state. The current `ETag` is used to check against the last updated time in the UTC of the file. If the date is different, it means that another activation of the same grain changed the state concurrently. In this situation, you'll throw an `InconsistentStateException`, which will result in the current activation being killed to prevent overwriting the state previously saved by the other activated grain.
```